Add rewrite base to generated .htaccess rules

IljaN opened the pull request:
<!--
Thanks for submitting a change to ownCloud!

This is the bug tracker for the Server component. Find other components at https://github.com/owncloud/core/blob/master/.github/CONTRIBUTING.md#guidelines

For fixing potential security issues please see https://owncloud.com/security/

To make it possible for us to get your change reviewed and merged please carefully fill out the requested information below.

Please note that any kind of change needs first be submitted to the master branch which holds the next version of ownCloud.

Please set the following labels:

- Set label "3 - To review" for review or "2 - Development" if the PR still has open tasks
- Assignment: assign to self
- Milestone: set the same as the ticket this PR fixes, or "development" by default
- Reviewers: pick at least one
-->

## Description
If htaccess.RewriteBase is set we need to prepend it in the generated rules.

## Related Issue
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->
- Fixes https://github.com/owncloud/core/issues/40696
- Related https://github.com/owncloud/core/pull/40584

## Motivation and Context
The [hardened .htaccess fix](https://github.com/owncloud/core/pull/40584) caused a regression for installations which have owncloud in a subfolder.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
- test environment:
- test case 1:
- test case 2:
- ...

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Database schema changes (next release will require increase of minor version instead of patch)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests only (no source changes)

## Checklist:
<!-- Tick the checkboxes when done. -->
<!-- Raise documentation ticket in https://github.com/owncloud/documentation -->
- [x] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link> 
- [ ] Changelog item, see [TEMPLATE](https://github.com/owncloud/core/blob/master/changelog/TEMPLATE)

IljaN committed:
Add rewrite base to generated .htaccess rules

If htaccess.RewriteBase is set we need to prepend it to the generated rules.

IljaN commented on lib/private/Setup.php:
```diff
@@ -491,26 +491,26 @@ public static function updateHtaccess() {
+			$content .= "\n  RewriteCond %{REQUEST_URI} !^$rewriteBase/ocm-provider/";
+			$content .= "\n  RewriteCond %{REQUEST_URI} !^$rewriteBase/\\.well-known/(acme-challenge|pki-validation)/.*";
 			$content .= "\n  RewriteRule . index.php [PT,E=PATH_INFO:$1]";
 			$content .= "\n  RewriteBase " . $rewriteBase;
```
Hmhh this would probably dont put a "/" here

ownclouders commented:
:boom: Acceptance tests pipeline <strong>apiSharePublicLink2-mariadb10.2-php7.4</strong> failed. The build has been cancelled.

https://drone.owncloud.com/owncloud/core/38112/60
<!-- id: pr-40697 -->

ownclouders commented:
:boom: Acceptance tests pipeline <strong>apiShareManagementBasicToShares-mariadb10.2-php7.4</strong> failed. The build has been cancelled.

https://drone.owncloud.com/owncloud/core/38112/54
<!-- id: pr-40697 -->

ownclouders commented:
:boom: Acceptance tests pipeline <strong>apiShareManagementToRoot-mariadb10.2-php7.4</strong> failed. The build has been cancelled.

https://drone.owncloud.com/owncloud/core/38112/51
<!-- id: pr-40697 -->

IljaN committed:
Fix potential double slash and empty rewriteBase if rewriteBase is "/"

jnweiger reviewed (changes requested):
A literal unescaped $rewriteBase can destroy the regexp. 
We need to apply proper regexp escaping, this is a security relevant.

IljaN commented:
We could do this in a addtional PR by sanitizing $rewriteBase with [preg_quote](https://www.php.net/manual/en/function.preg-quote.php). But IMO offtopic for this PR. A path shouldn't contain any Regex characters and is coming from the trusted config anyways, no?

IljaN commented:
Fail: https://drone.owncloud.com/owncloud/core/38114/172/15

jnweiger committed:
use preg_quote() for proper regex escaping.

jnweiger commented:
> We could do this in a addtional PR by sanitizing $rewriteBase with [preg_quote](https://www.php.net/manual/en/function.preg-quote.php). But IMO offtopic for this PR. A path shouldn't contain any Regex characters and is coming from the trusted config anyways, no?

No, I don't think so. This path is perfectly legal, imho.
occ config:system:set htaccess.RewriteBase --value '/funny$nameing-scheme-*-but+we+like+it'


Pfft. On another thought, we don't quote strings in .htaccess, that means it breaks apart, when RewriteBase contains whitespace...

jnweiger commented on lib/private/Setup.php:
```diff
@@ -492,25 +492,31 @@ public static function updateHtaccess() {
 		// Add rewrite rules if the RewriteBase is configured
 		$rewriteBase = $config->getSystemValue('htaccess.RewriteBase', '');
 		if ($rewriteBase !== '') {
+			$rewriteBase_re = \preg_quote($rewriteBase_re);
```
```suggestion
			$rewriteBase_re = \preg_quote($rewriteBase);
```

IljaN replied:
Okay definetly makes sense for "-" characters. Thx

jnweiger committed:
Update lib/private/Setup.php

ho4ho commented:
Fix #40686 also

jnweiger committed:
@IljaN we have a double // in the way, the test suite generates the rewriteBase.
I am adding a trim, to get rid of double slashes. That should make the CI pass.

ho4ho commented on lib/private/Setup.php:
```diff
@@ -492,25 +492,30 @@ public static function updateHtaccess() {
+			$content .= "\n  RewriteCond %{REQUEST_URI} !^$rewriteBaseRe/updater/";
+			$content .= "\n  RewriteCond %{REQUEST_URI} !^$rewriteBaseRe/ocs-provider/";
+			$content .= "\n  RewriteCond %{REQUEST_URI} !^$rewriteBaseRe/ocm-provider/";
+			$content .= "\n  RewriteCond %{REQUEST_URI} !^$rewriteBaseRe/\\.well-known/(acme-challenge|pki-validation)/.*";
```
```suggestion
			$content .= "\n  RewriteCond %{REQUEST_URI} !^/\\.well-known/(acme-challenge|pki-validation)/.*";
```

Thinking always needing at root and not by subfolder.
